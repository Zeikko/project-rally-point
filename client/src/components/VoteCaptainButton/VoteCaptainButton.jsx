import _ from 'lodash'
import PropTypes from 'prop-types'
import React, { Component } from 'react'
import { voteCaptainAction } from '../../actions/captain-vote-actions'
import Button from '../Button/Button'
import * as propTypes from '../../constants/prop-types'
import gameStatuses from '../../../../common/game-statuses.json'

class VoteCaptainButton extends Component {
  constructor() {
    super()
    this.handleClick = this.handleClick.bind(this)
  }

  handleClick() {
    const { dispatch, player, gameState } = this.props
    dispatch(voteCaptainAction(gameState.game.id, player.id))
  }

  render() {
    const {
      player, gameState, captainVotesState, userState,
    } = this.props
    if (gameState.game.status !== gameStatuses.VOTE_CAPTAINS) {
      return null
    }
    const userHasVoted = _.find(captainVotesState.captainVotes, { voterId: userState.user.id })
    if (userHasVoted) {
      const votes = _.filter(captainVotesState.captainVotes, { votedId: player.id }).length
      return <span>{ votes }</span>
    }
    return (
      <Button onClick={this.handleClick}>
        Vote
      </Button>
    )
  }
}

VoteCaptainButton.propTypes = {
  dispatch: PropTypes.func.isRequired,
  player: propTypes.user.isRequired,
  gameState: propTypes.gameState.isRequired,
  captainVotesState: propTypes.captainVotesState.isRequired,
  userState: propTypes.userState.isRequired,
}

export default VoteCaptainButton
