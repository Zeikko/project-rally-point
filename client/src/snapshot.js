import renderer from 'react-test-renderer'
import 'jest-styled-components'

export function matchComponentToSnapshot(component) {
  const json = renderer.create(component).toJSON()
  expect(json).toMatchSnapshot()
}