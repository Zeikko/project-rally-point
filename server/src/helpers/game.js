import _ from 'lodash'
import db from '../db'
import gameStatuses from '../../../common/game-statuses.json'

export async function shouldStartCaptainVote(gameId) {
  const game = await db('game')
    .first('*')
    .where({ id: gameId })
  const numberOfPlayers = await db('player')
    .count('id')
    .where({ gameId })
  return parseInt(numberOfPlayers[0].count, 10) === game.maxPlayers
}

export async function shouldStartSquadLeaderPick(gameId) {
  const game = await db('game')
    .first('*')
    .where({ id: gameId })
  const numberOfCaptainVotes = await db('captainVote')
    .count('id')
    .where({ gameId })
  return parseInt(numberOfCaptainVotes[0].count, 10) === game.maxPlayers
}

export async function startCaptainVote(gameId) {
  const games = await db('game')
    .update({ status: gameStatuses.VOTE_CAPTAINS })
    .where({ id: gameId })
    .returning('*')
  return _.first(games)
}

export async function startSquadLeaderPick(gameId) {
  const games = await db('game')
    .update({ status: gameStatuses.SQUAD_LEADER_PICK })
    .where({ id: gameId })
    .returning('*')
  return _.first(games)
}

export function createNewGame() {
  return db('game').insert({})
}

export function getGameStatus(gameId) {
  return db('game')
    .first('status')
    .where({ id: gameId })
}
