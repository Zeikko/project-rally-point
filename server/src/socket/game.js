import logger from '../logger'
import db from '../db'
import actions from '../../../common/actions.json'

export default function handleGame(io, socket, action) {
  switch (action.type) {
    case actions.GET_GAME_REQUEST:
      return getGameRequest(io, socket)
    case actions.CREATE_GAME_REQUEST:
      return createGameRequest(io, socket)
    default:
      return null
  }
}

export async function getGameRequest(io, socket) {
  try {
    const game = await db('game')
      .orderBy('id', 'desc')
      .first('*')
    socket.emit('action', { type: actions.GET_GAME_SUCCESS, data: game })
    return game
  } catch (error) {
    logger.error(error)
    socket.emit('action', { type: actions.GET_GAME_ERROR })
  }
  return null
}

async function createGameRequest(io, socket) {
  try {
    const { 0: game } = await db('game')
      .insert({})
      .returning('*')
    socket.emit('action', { type: actions.GET_GAME_SUCCESS, data: game })
  } catch (error) {
    logger.error(error)
    socket.emit('action', { type: actions.GET_GAME_ERROR })
  }
}
