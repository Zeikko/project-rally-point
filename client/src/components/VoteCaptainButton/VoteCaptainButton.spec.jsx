import React from 'react'
import { mount } from 'enzyme'
import { matchComponentToSnapshot } from '../../test/snapshot'
import VoteCaptainButton from './VoteCaptainButton'
import game from '../../../../fixtures/game.json'
import normalUser from '../../../../fixtures/normal-user.json'
import initialState from '../../initial-state'
import gameStatuses from '../../../../common/game-statuses.json'

describe('VoteCaptainButton', () => {
  it('does not render when not voting for a captain', () => {
    matchComponentToSnapshot(<VoteCaptainButton
      dispatch={() => {}}
      player={normalUser}
      gameState={{ game }}
      captainVotesState={initialState.captainVotesState}
      userState={initialState.userState}
    />)
  })

  it('renders button when voting for a captain', () => {
    matchComponentToSnapshot(<VoteCaptainButton
      dispatch={() => {}}
      player={normalUser}
      gameState={{ game: { ...game, status: gameStatuses.VOTE_CAPTAINS } }}
      captainVotesState={initialState.captainVotesState}
      userState={{ isLoading: false, user: normalUser }}
    />)
  })

  it('votes a captain', () => {
    const dispatch = jest.fn()
    const wrapper = mount(<VoteCaptainButton
      dispatch={() => {}}
      player={normalUser}
      gameState={{ game: { ...game, status: gameStatuses.VOTE_CAPTAINS } }}
      captainVotesState={initialState.captainVotesState}
      userState={{ isLoading: false, user: normalUser }}
    />)
    wrapper.find('button').simulate('click')
    expect(dispatch.mock.calls[0]).toEqual([
      {
        type: 'VOTE_CAPTAIN_REQUEST',
        socket: true,
        gameId: 1,
      },
    ])
  })
})
