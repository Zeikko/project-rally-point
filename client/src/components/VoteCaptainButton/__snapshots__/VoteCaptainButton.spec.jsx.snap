// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VoteCaptainButton does not render when not voting for a captain 1`] = `null`;

exports[`VoteCaptainButton renders button when voting for a captain 1`] = `
.c0 {
  font-size: 18px;
  border: solid black 1px;
  background-color: #AAA;
  cursor: pointer;
}

<button
  className="c0"
  onClick={[Function]}
>
  Vote
</button>
`;
